// @matt-dunn/react-wireframes/lib/withWireframeAnnotation/Identifier.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

require("core-js/modules/es.array.slice");

require("core-js/modules/es.object.freeze");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Identifier = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styled = _interopRequireDefault(require("@emotion/styled"));

var _reactTransitionGroup = require("react-transition-group");

var _Identifier = require("../components/Identifier");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  position: absolute;\n  top: -1em;\n  left: -1em;\n  z-index: 4000;\n  cursor: default;\n  transition: opacity 250ms, visibility 250ms;\n  opacity: 0.75;\n  visibility: visible;\n\n  &.fade-Identifier.enter {\n    opacity: 0;\n    visibility: hidden;\n  }\n  &.fade-Identifier.enter-active {\n    opacity: 0.75;\n    visibility: visible;\n  }\n  &.fade-Identifier.exit {\n    opacity: 0.75;\n    visibility: visible;\n  }\n  &.fade-Identifier.exit-active {\n    opacity: 0;\n    visibility: hidden;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var IdentifierContainer = (0, _styled.default)(_Identifier.Identifier)(_templateObject());

var Identifier = function Identifier(_ref) {
  var annotation = _ref.annotation,
      parentReference = _ref.parentReference,
      _ref$show = _ref.show,
      show = _ref$show === void 0 ? true : _ref$show;
  var nodeRef = (0, _react.useRef)(null);
  return /*#__PURE__*/_react.default.createElement(_reactTransitionGroup.CSSTransition, {
    nodeRef: nodeRef,
    timeout: 250,
    className: "fade-Identifier",
    in: show,
    mountOnEnter: true,
    unmountOnExit: true
  }, /*#__PURE__*/_react.default.createElement(IdentifierContainer, {
    ref: nodeRef,
    "data-annotation-identifier": true,
    annotation: annotation,
    parentReference: parentReference
  }));
};

exports.Identifier = Identifier;
