// @matt-dunn/react-wireframes/lib/withWireframeAnnotation/withWireframeAnnotation.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

require("core-js/modules/es.symbol");

require("core-js/modules/es.symbol.description");

require("core-js/modules/es.symbol.iterator");

require("core-js/modules/es.array.from");

require("core-js/modules/es.array.index-of");

require("core-js/modules/es.array.iterator");

require("core-js/modules/es.array.slice");

require("core-js/modules/es.function.name");

require("core-js/modules/es.object.freeze");

require("core-js/modules/es.object.keys");

require("core-js/modules/es.object.to-string");

require("core-js/modules/es.regexp.to-string");

require("core-js/modules/es.string.iterator");

require("core-js/modules/web.dom-collections.iterator");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withWireframeAnnotation = withWireframeAnnotation;
exports.WireframeAnnotationPropsInterfaceDefinition = exports.withWireframeAnnotationInterfaceDefinition = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styled = _interopRequireDefault(require("@emotion/styled"));

var _css = _interopRequireDefault(require("@emotion/css"));

var _classnames = _interopRequireDefault(require("classnames"));

var _useApi2 = require("../useApi");

var _WireframeProvider = require("../WireframeProvider");

var _Identifier = require("./Identifier");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n      z-index: 5000;\n      \n      > * {\n        box-shadow: 0 0 0 1px #4086f7 !important;\n      }\n    "]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  position: relative;\n  \n  [disabled] {\n    pointer-events: none;\n  }\n  \n  &.outlined,\n  &.outline:hover {\n    ", "\n  }\n\n  &.outlined,\n  &:hover {\n    > [data-annotation-identifier] {\n        transition: opacity 0ms, visibility 0ms;\n        opacity: 1;\n    }\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var Wrapper = _styled.default.span(_templateObject(), function (_ref) {
  var show = _ref.show;
  return show && (0, _css.default)(_templateObject2());
});
/* istanbul ignore next */


var getDisplayName = function getDisplayName(WrappedComponent) {
  return typeof WrappedComponent === "string" ? WrappedComponent : WrappedComponent.displayName || WrappedComponent.name || "Component";
};

function withWireframeAnnotation(options) {
  return function withWireframeAnnotationComponent(WrappedComponent) {
    var Component = /*#__PURE__*/_react.default.memo(function (props) {
      return /*#__PURE__*/_react.default.createElement(WrappedComponent, props);
    });

    Component.displayName = "withWireframeAnnotation(".concat(getDisplayName(WrappedComponent), ")"); // eslint-disable-next-line react/no-multi-comp

    function WrappedWireframeAnnotation(_ref2) {
      var className = _ref2.className,
          _ref2$outline = _ref2.outline,
          outline = _ref2$outline === void 0 ? true : _ref2$outline,
          _ref2$isHighlighted = _ref2.isHighlighted,
          isHighlighted = _ref2$isHighlighted === void 0 ? false : _ref2$isHighlighted,
          props = _objectWithoutProperties(_ref2, ["className", "outline", "isHighlighted"]);

      var _useApi = (0, _useApi2.useApi)(),
          register = _useApi.register,
          unregister = _useApi.unregister,
          onOpen = _useApi.onOpen,
          isOpen = _useApi.isOpen,
          highlightNote = _useApi.highlightNote,
          getParentReference = _useApi.getParentReference;

      var _useState = (0, _react.useState)(),
          _useState2 = _slicedToArray(_useState, 2),
          annotation = _useState2[0],
          setAnnotation = _useState2[1];

      var _useState3 = (0, _react.useState)(isOpen()),
          _useState4 = _slicedToArray(_useState3, 2),
          show = _useState4[0],
          setShow = _useState4[1];

      (0, _react.useEffect)(function () {
        var cb = onOpen(setShow);
        setAnnotation(register(Component, options));
        return function () {
          cb.unregister();
          unregister(Component);
        };
      }, []); // eslint-disable-line react-hooks/exhaustive-deps

      var handleHighlightNote = (0, _react.useCallback)(function (e) {
        e.stopPropagation();
        /* istanbul ignore else */

        if (show) {
          highlightNote(Component);
        }
      }, [show, highlightNote]);
      var handleHighlightNoteReset = (0, _react.useCallback)(function () {
        /* istanbul ignore else */
        if (show) {
          highlightNote(undefined);
        }
      }, [show, highlightNote]);
      return /*#__PURE__*/_react.default.createElement(_WireframeProvider.WireframeAnnotationComponentContext.Provider, {
        value: annotation
      }, /*#__PURE__*/_react.default.createElement(Wrapper, {
        show: show,
        onMouseMove: handleHighlightNote,
        onFocus: handleHighlightNote,
        onMouseLeave: handleHighlightNoteReset,
        className: (0, _classnames.default)({
          "outline": outline,
          "outlined": isHighlighted && show
        }, className)
      }, annotation && /*#__PURE__*/_react.default.createElement(_Identifier.Identifier, {
        annotation: annotation,
        parentReference: getParentReference(),
        show: show
      }), /*#__PURE__*/_react.default.createElement(Component, props)));
    }

    WrappedWireframeAnnotation.Component = Component;
    return WrappedWireframeAnnotation;
  };
}
/* istanbul ignore next */


var withWireframeAnnotationInterfaceDefinition = function withWireframeAnnotationInterfaceDefinition(params) {
  return null;
};
/* istanbul ignore next */


exports.withWireframeAnnotationInterfaceDefinition = withWireframeAnnotationInterfaceDefinition;

var WireframeAnnotationPropsInterfaceDefinition = function WireframeAnnotationPropsInterfaceDefinition(_ref3) {
  var _ref3$outline = _ref3.outline,
      outline = _ref3$outline === void 0 ? true : _ref3$outline,
      _ref3$isHighlighted = _ref3.isHighlighted,
      isHighlighted = _ref3$isHighlighted === void 0 ? false : _ref3$isHighlighted;
  return null;
};

exports.WireframeAnnotationPropsInterfaceDefinition = WireframeAnnotationPropsInterfaceDefinition;
